File: two-sum-ii-input-array-is-sorted.py
class Solution:
    # two pointers: O(n)
    def twoSum(self, numbers: List[int], target: int) -> List[int]:
        left, right = 0, len(numbers) - 1
        while left != right:
            if numbers[left] + numbers[right] < target:
                left += 1
            elif numbers[left] + numbers[right] > target:
                right -= 1
            else:
                return [left + 1, right + 1]

    # binary search without slicing: O(nlogn)
    def twoSum(self, numbers: List[int], target: int) -> List[int]:
        k = 0
        while k < len(numbers): # O(n)
            number = numbers[k]
            index = bisect.bisect_left(numbers, target - number, k + 1) # O(logn)
            if index < len(numbers) and numbers[index] == target - number:
                return [k + 1, index + 1]
            k += 1

    # binary search with slicing: O(nlogn)
    def twoSum(self, numbers: List[int], target: int) -> List[int]:
        k = 0
        while k < len(numbers): # O(n)
            number = numbers[k]
            index = bisect.bisect_left(numbers[k+1:], target - number) # O(logn)
            if k+1+index < len(numbers) and numbers[k+1+index] == target - number:
                return [k + 1, k + 1 + index + 1]
            k += 1

File: two-sum-ii-input-array-is-sorted.java
class Solution {
    public int[] twoSum(int[] numbers, int target) {
        for(int i=0; i<numbers.length - 1; i++) {
            int t = target - numbers[i];
            int left = 0, right = numbers.length -1;
            int mid = -1;
            while(left <= right) {
                mid = left + (right - left) / 2;
                if(numbers[mid] == t) {
                    if(mid != i)
                        return new int[]{i + 1, mid + 1};
                    else if(numbers[mid] == numbers[mid + 1])
                        return new int[]{i + 1, mid + 2};
                }
                else if(numbers[mid] < t) {
                    left = mid + 1;
                } else {
                    right = mid -1;
                }
            }
        }
        return new int[]{};
    }
}